WcaOnRails/app/models/person.rb

```diff
@@ -1,13 +1,93 @@
 class Person < ActiveRecord::Base
-  self.table_name = "Persons"
+  self.table_name = "rails_persons"
   self.primary_key = "id"
-  has_one :user, foreign_key: "wca_id"
-  has_many :results, foreign_key: "personId"
+
+  has_one :user, primary_key: "wca_id", foreign_key: "wca_id"
+  has_many :results, primary_key: "wca_id", foreign_key: "personId"
   has_many :competitions, -> { distinct }, through: :results
-  has_many :ranksAverage, foreign_key: "personId", class_name: "RanksAverage"
-  has_many :ranksSingle, foreign_key: "personId", class_name: "RanksSingle"
+  has_many :ranksAverage, primary_key: "wca_id", foreign_key: "personId", class_name: "RanksAverage"
+  has_many :ranksSingle, primary_key: "wca_id", foreign_key: "personId", class_name: "RanksSingle"
+
+  scope :current, -> { where(subId: 1) }
+
+  validates :name, presence: true
+  validates :countryId, presence: true
+
+  before_validation :unpack_dob
+  private def unpack_dob
+    if @dob.nil? && !dob.blank?
+      @dob = dob.strftime("%F")
+    end
+    if @dob.blank?
+      self.year = self.month = self.day = 0
+    else
+      unless @dob =~ /\A\d{4}-\d{2}-\d{2}\z/
+        errors.add(:dob, "is invalid")
+        return false
+      end
+      self.year, self.month, self.day = @dob.split("-").map(&:to_i)
+      unless Date.valid_date? self.year, self.month, self.day
+        errors.add(:dob, "is invalid")
+        return false
+      end
+    end
+  end
 
-  alias_method :wca_id, :id
+  validate :dob_must_be_in_the_past
+  private def dob_must_be_in_the_past
+    if dob && dob >= Date.today
+      errors.add(:dob, "must be in the past")
+    end
+  end
+
+  # If someone represented country A, and now represents country B, it's
+  # easy to tell which solves are which (results have a countryId).
+  # Fixing their country (B) to a new country C is easy to undo, just change
+  # all Cs to Bs. However, if someone accidentally fixes their country from B
+  # to A, then we cannot go back, as all their results are now for country A.
+  validate :cannot_change_country_to_country_represented_before, if: :countryId_changed?, unless: -> { new_record? || @updating_using_sub_id }
+  private def cannot_change_country_to_country_represented_before
+    has_represented_this_country_already = Person.exists?(wca_id: wca_id, countryId: countryId)
+    if has_represented_this_country_already
+      errors.add(:countryId, "Cannot change the country to a country the person have already represented in the past.")
```

Comment:
nit: I see you changed this message from "Cannot change the country of someone to a country they have already represented in the past.". I'm ok with your change, but the new message is a bit grammatically incorrect. Try changing have -> has:

> Cannot change the country to a country the person has already represented in the past.